### ISSUE TITLE
[callback-return] with optional callback pattern

### ISSUE THREAD

**Issue by galambalazs** [eslint/eslint#5599]
There's the simple case when the callback is optional so given a condition you might want to exit early, but only call the `callback` if it's defined. Consider this example:

```javascript
function doSomething(callback) {
  if (alreadyDone) {
    if (callback) callback(null, 'ok'); 
    return;
  }
  // .. do something
  callback(null, 'ok');
}
```

This warns because `callback` is used without return even though it's on the next statement.

**Comment by platinumazure** [comment]
Just out of curiosity, why not `if (callback) return callback(null, 'ok')`?

**Comment by galambalazs** [comment]
Because you want to return based on the condition `alreadyDone`
(that's independent of the callback being defined or not)

**Comment by platinumazure** [comment]
Please flesh out your proposal a little more. Can you please provide a few more examples of when the rule should not warn but currently does, or when the rule should warn but currently doesn't?

**Comment by galambalazs** [comment]
Neither of these should throw an error:

1: return with the callback
```
    return callback(null, 'ok');
```

2: return is the first statement after callback
```
callback(null, 'ok');
return;
```

3: return is the first statement after the callback's {block}
```
if (CONDITION) { 
  callback(null, 'ok');
}
return;
```

Currently (3) throws an error.

**Comment by platinumazure** [comment]
Okay, if your second example already does not warn, I could get behind allowing your third example. With our code path analysis tools now available, I think it could be pretty easy (but I could be wrong).

@mysticatea I would be grateful if you would read the comment above mine and my analysis and post whether or not my analysis makes sense. Thanks!

**Comment by galambalazs** [comment]
Yes, thank you. I was just being explicit, it's true that the problem is currently the (3) case.
(that's why I gave the title to this issue: *optional callback pattern*)

Please note, that I'm aware that there are other ways to write that conditional like:

```
return callback && callback(null, 'ok');
```

But this doesn't mean that the form I showed above as (3) should throw an error. It is perfectly valid code  and it's very easy to see that it does return after the (optional) callback is called.

**Comment by platinumazure** [comment]
@galambalazs What about if the callback is a few blocks deep?

```js
function f(data, cb) {
    if (data) {
        if (crunchData()) {
            if (cb) {
                cb(null, data);
            }
        }
        return;
    }
}

**Comment by galambalazs** [comment]
I think those cases are harder to handle. I'm not sure what the capabilities of eslint are.

First I would concentrate on the very simple case that's currently failing.
(when I cannot say I just return with the results of the callback, because the callback may not be defined, so first I'll have to check if it's safe to call it)

As for your example, personally I don't being explicit in those cases with multiple returns until eslint can better analyze the code.

**Comment by kaicataldo** [comment]
Closing this because we haven't made any progress in months. Thanks for contributing!
